cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -585,6 +586,37 @@ public Set<ProgramId> filterProgramsExistence(Collection<ProgramReference> progr
         .collect(Collectors.toSet());
   }
 
+  /**
+   * Marks a given application version as latest. This also unmarks the previous latest version.
+   *
+   * @param id {@link ApplicationId}
+   * @throws IOException when the updates fail
+   * @throws ApplicationNotFoundException when the application is not found
+   */
+  public void updateLatestApplicationVersion(ApplicationId id)
+    throws IOException, ApplicationNotFoundException {
+    StructuredTable appSpecTable = getApplicationSpecificationTable();
+
+    // First unmark the current latest version
+    ApplicationMeta latest = getLatest(id.getAppReference());
```

Comment:
getLatest reads the entire row and deserializes into ApplicationMeta, which is more expensive than needed. It also has a bunch of extra reads to handle rows that were in the table before the isLatest column was added.

For this method, none of that extra logic is needed. We just need to use the StructuredTable.update() method to first set whatever has isLatest == true to false, and then to set isLatest to true for the id given.

Reply:
for the existence check, I don't remember if trying to update a row that doesn't exist ends up throwing an exception or if it becomes a no-op. If it's a no-op and we still need to do the read, we should be using StructuredTable.read() on just the relevant fields without deserializing anything into an ApplicationMeta.

Reply:
the update method does not throw an exception when the row does not exist. It's a no-op in that case. 

Updated the code to read the application row to check if it exists already.